File: convert-consultations.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'fileutils'
require 'date'

##
# Conversion script to transform legacy Ayurvedic consultation markdown files
# to structured JSON format with standardized recommendations.
class ConsultationParser
  def initialize
    # Load standard recommendations
    standard_recs_file = File.read('standard-recommendations.json')
    @standard_recs = JSON.parse(standard_recs_file)['standardRecommendations']
  end

  def parse_patient_info(content)
    patient = {}

    # Get first 10 lines for basic info
    lines = content.split("\n")[0..9]

    lines.each do |line|
      cleaned_line = line.gsub(/\*/, '').strip

      # Name
      if cleaned_line.include?('Name') && cleaned_line.include?('–')
        name_match = cleaned_line.match(/Name\s*[-–]\s*([^–]+?)(?:\s+Date|$)/)
        patient['name'] = name_match[1].strip if name_match
      end

      # Date
      if cleaned_line.include?('Date') && cleaned_line.include?('–')
        date_match = cleaned_line.match(/Date\s*[-–]\s*([^–]+?)(?:\s+|$)/)
        patient['date'] = date_match[1].strip if date_match
      end

      # Age
      if cleaned_line.include?('Age') && cleaned_line.include?('years')
        age_match = cleaned_line.match(/Age\s*[-–]\s*(\d+)\s*years/)
        patient['age'] = age_match[1].to_i if age_match
      end

      # Sex
      if cleaned_line.include?('Sex') && cleaned_line.include?('–')
        sex_match = cleaned_line.match(/Sex\s*[-–]\s*([^–]+?)(?:\s+|$)/)
        patient['sex'] = sex_match[1].strip if sex_match
      end

      # Profession
      if cleaned_line.include?('Profession') && cleaned_line.include?('–')
        prof_match = cleaned_line.match(/Profession\s*[-–]\s*([^–]+?)(?:\s+|$)/)
        patient['profession'] = prof_match[1].strip if prof_match
      end
    end

    # Family information
    family = {}

    family_text = extract_section(content, /Family.*?(?=Married|Present|$)/m)
    family['hasFamily'] = family_text&.include?('Yes') || false

    married_text = extract_section(content, /Married.*?(?=Children|Present|$)/m)
    if married_text
      family['married'] = married_text.gsub('Married –', '').strip
    end

    children_text = extract_section(content, /Children.*?(?=Born|Present|$)/m)
    if children_text
      family['children'] = children_text.gsub('Children –', '').strip
    end

    patient['family'] = family unless family.empty?

    # Birth place
    birth_match = content.match(/Born in\s*[-–]?\s*([^\n]+)/)
    patient['birthPlace'] = birth_match[1].strip if birth_match

    # Pets
    pets_match = content.match(/Pets\s*[-–]\s*([^\n]+)/)
    patient['pets'] = pets_match[1].strip if pets_match

    # Email
    email_match = content.match(/Email\s*[-–:]\s*([^\s\n]+@[^\s\n]+)/)
    patient['email'] = email_match[1].strip if email_match

    # Phone
    phone_match = content.match(/Phone\s*[-–:]\s*([^\n]+)/)
    patient['phone'] = phone_match[1].strip if phone_match

    patient
  end

  def extract_section(content, pattern)
    match = content.match(pattern)
    match ? match[0] : nil
  end

  def parse_consultation_info(content)
    consultation = {}

    # Present complaint
    complaint_match = content.match(/Present complaint\*\*\s*\n(.*?)(?=\*\*|No Covid|TM|$)/m)
    consultation['presentComplaint'] = clean_text(complaint_match[1]) if complaint_match

    # COVID history
    covid_matches = content.scan(/(No Covid[^*\n]*|Covid positive[^*\n]*|No vaccine[^*\n]*|vaccine[^*\n]*)/i)
    unless covid_matches.empty?
      consultation['covidHistory'] = covid_matches.flatten.join(', ')
    end

    # TM practice
    tm_match = content.match(/TM.*?(?=\n|$)/)
    consultation['tmPractice'] = tm_match[0] if tm_match

    consultation
  end

  def parse_assessment(content)
    assessment = {
      'vitals' => {},
      'lifestyle' => { 'food' => {} },
      'ayurvedicAssessment' => {}
    }

    # Vitals patterns
    vitals_patterns = {
      'pulse' => /Pulse\s*[-–]\s*([^\n]+)/i,
      'bowel' => /Bowel\s*[-–]\s*([^\n]+)/i,
      'urination' => /Urination\s*[-–]\s*([^\n]+)/i,
      'tongue' => /Tongue\s*[-–]\s*([^\n]+)/i,
      'sleep' => /Sleep\s*[-–]\s*([^\n]+)/i,
      'hunger' => /Hunger\s*[-–]\s*([^\n]+)/i,
      'thirst' => /Thirst\s*[-–]\s*([^\n]+)/i,
      'menstruation' => /Menstruation\s*[-–]\s*([^\n]+)/i
    }

    vitals_patterns.each do |key, pattern|
      match = content.match(pattern)
      assessment['vitals'][key] = match[1].strip if match
    end

    # Lifestyle patterns
    lifestyle_patterns = {
      'dailyRoutine' => /Daily routine\s*[-–]\s*([^\n]+)/i,
      'smoking' => /Smoking\s*[-–]\s*([^\n]+)/i,
      'alcohol' => /Alcohol\s*[-–]\s*([^\n]+)/i,
      'exercise' => /Exercise\s*[-–]\s*([^\n]+)/i,
      'emotions' => /Emotions\s*[-–]\s*([^\n]+)/i
    }

    lifestyle_patterns.each do |key, pattern|
      match = content.match(pattern)
      assessment['lifestyle'][key] = match[1].strip if match
    end

    # Food patterns
    food_patterns = {
      'hoteling' => /Food\s*[-–]\s*([^\n]+)/i,
      'oils' => /Oil\s*[-–]\s*([^\n]+)/i,
      'breakfast' => /Breakfast\s*[-–]\s*([^\n]+)/i,
      'lunch' => /Lunch\s*[-–]\s*([^\n]+)/i,
      'dinner' => /Dinner\s*[-–]\s*([^\n]+)/i,
      'fruits' => /Fruits\s*[-–]\s*([^\n]+)/i
    }

    food_patterns.each do |key, pattern|
      match = content.match(pattern)
      assessment['lifestyle']['food'][key] = match[1].strip if match
    end

    # Ayurvedic Assessment patterns
    ayur_patterns = {
      'prakruti' => /Prakruti\s*[-–]\s*([^\n]+)/i,
      'vikruti' => /Vikruti\s*[-–]\s*([^\n]+)/i,
      'dosha' => /Dosha\s*[-–]\s*([^\n]+)/i,
      'dushya' => /Dushya\s*[-–]\s*([^\n]+)/i
    }

    ayur_patterns.each do |key, pattern|
      match = content.match(pattern)
      assessment['ayurvedicAssessment'][key] = match[1].strip if match
    end

    assessment
  end

  def parse_treatment(content)
    treatment = {
      'shamana' => [],
      'panchakarma' => [],
      'additionalAdvice' => []
    }

    # Shamana (Internal Medicine)
    shamana_section = extract_section(content, /Shamana.*?(?=Panchakarma|Advice|Avoid)/m)
    if shamana_section
      medicine_lines = shamana_section.scan(/(\d+)\.\s*\*\*(.*?)\*\*/)
      medicine_lines.each do |_, medicine_text|
        if medicine_text.include?('–')
          parts = medicine_text.split('–', 2)
          medicine_name = parts[0].strip
          dosage = parts.length > 1 ? parts[1].strip : ""
          treatment['shamana'] << {
            'medicine' => medicine_name,
            'dosage' => dosage
          }
        end
      end
    end

    # Panchakarma
    panchakarma_section = extract_section(content, /Panchakarma\*\*(.*?)(?=\*\*Advice|\*\*Avoid)/m)
    if panchakarma_section
      panchakarma_lines = panchakarma_section.scan(/(\d+)\.\s*\*\*(.*?)\*\*/)
      panchakarma_lines.each do |_, therapy|
        treatment['panchakarma'] << therapy.strip
      end
    end

    # Additional Advice
    advice_patterns = [
      /Aroma therapy[^\n]*/i,
      /Chanting of mantras[^\n]*/i,
      /Advice to do Moolabandha/i,
      /Use of Asafetida[^\n]*/i,
      /Drink medicated water[^\n]*/i
    ]

    advice_patterns.each do |pattern|
      matches = content.scan(pattern)
      treatment['additionalAdvice'].concat(matches.flatten)
    end

    treatment
  end

  def clean_text(text)
    return '' unless text

    # Remove markdown formatting
    text = text.gsub(/\*\*/, '')
    # Remove extra whitespace
    text = text.gsub(/\s+/, ' ')
    text.strip
  end

  def parse_file(filepath)
    content = File.read(filepath, encoding: 'utf-8')

    consultation_data = {
      'patient' => parse_patient_info(content),
      'consultation' => parse_consultation_info(content),
      'assessment' => parse_assessment(content),
      'treatment' => parse_treatment(content),
      'standardRecommendations' => @standard_recs,
      'metadata' => {
        'createdAt' => DateTime.now.iso8601,
        'lastModified' => DateTime.now.iso8601,
        'version' => '1.0',
        'language' => 'en',
        'originalFile' => File.basename(filepath)
      }
    }

    consultation_data
  end

  def convert_all_files(input_dir: '.', output_dir: './converted')
    FileUtils.mkdir_p(output_dir)

    # Find all consultation files
    consultation_files = Dir.entries(input_dir).select do |f|
      f.end_with?('.md') &&
        !f.downcase.include?('consultation') &&
        !f.downcase.include?('template') &&
        !f.downcase.include?('readme')
    end

    puts "Found #{consultation_files.length} consultation files to convert"

    consultation_files.each do |filename|
      begin
        puts "Converting #{filename}..."
        filepath = File.join(input_dir, filename)
        consultation_data = parse_file(filepath)

        # Create output filename
        base_name = File.basename(filename, '.md')
        output_filename = "#{base_name}.json"
        output_path = File.join(output_dir, output_filename)

        # Save JSON
        File.write(output_path, JSON.pretty_generate(consultation_data))

        puts "✓ Converted to #{output_filename}"

      rescue => e
        puts "✗ Error converting #{filename}: #{e.message}"
      end
    end

    puts "\nConversion complete! Check the '#{output_dir}' directory for results."
  end
end

# Main execution
if __FILE__ == $0
  parser = ConsultationParser.new
  parser.convert_all_files
end

File: convert-consultations.py
#!/usr/bin/env python3
"""
Conversion script to transform legacy Ayurvedic consultation markdown files
to structured JSON format with standardized recommendations.
"""

import re
import json
import os
from datetime import datetime
from typing import Dict, List, Any, Optional

class ConsultationParser:
    def __init__(self):
        # Load standard recommendations
        with open('standard-recommendations.json', 'r', encoding='utf-8') as f:
            self.standard_recs = json.load(f)['standardRecommendations']

    def parse_patient_info(self, content: str) -> Dict[str, Any]:
        """Extract patient demographic information"""
        patient = {}

        # Name and basic info - first few lines
        lines = content.split('\n')[:10]

        for line in lines:
            line = line.replace('*', '').strip()

            # Name
            if 'Name' in line and '–' in line:
                name_match = re.search(r'Name\s*[-–]\s*([^–]+?)(?:\s+Date|$)', line)
                if name_match:
                    patient['name'] = name_match.group(1).strip()

            # Date
            if 'Date' in line and '–' in line:
                date_match = re.search(r'Date\s*[-–]\s*([^–]+?)(?:\s+|$)', line)
                if date_match:
                    patient['date'] = date_match.group(1).strip()

            # Age
            if 'Age' in line and 'years' in line:
                age_match = re.search(r'Age\s*[-–]\s*(\d+)\s*years', line)
                if age_match:
                    patient['age'] = int(age_match.group(1))

            # Sex
            if 'Sex' in line and '–' in line:
                sex_match = re.search(r'Sex\s*[-–]\s*([^–]+?)(?:\s+|$)', line)
                if sex_match:
                    patient['sex'] = sex_match.group(1).strip()

            # Profession
            if 'Profession' in line and '–' in line:
                prof_match = re.search(r'Profession\s*[-–]\s*([^–]+?)(?:\s+|$)', line)
                if prof_match:
                    patient['profession'] = prof_match.group(1).strip()

        # Family information
        family = {}
        family_text = self.extract_section(content, r'Family.*?(?=Married|Present|$)', multiline=True)
        if family_text:
            family['hasFamily'] = 'Yes' in family_text

        married_text = self.extract_section(content, r'Married.*?(?=Children|Present|$)', multiline=True)
        if married_text:
            family['married'] = married_text.replace('Married –', '').strip()

        children_text = self.extract_section(content, r'Children.*?(?=Born|Present|$)', multiline=True)
        if children_text:
            family['children'] = children_text.replace('Children –', '').strip()

        if family:
            patient['family'] = family

        # Birth place
        birth_match = re.search(r'Born in\s*[-–]?\s*([^\n]+)', content)
        if birth_match:
            patient['birthPlace'] = birth_match.group(1).strip()

        # Pets
        pets_match = re.search(r'Pets\s*[-–]\s*([^\n]+)', content)
        if pets_match:
            patient['pets'] = pets_match.group(1).strip()

        # Email
        email_match = re.search(r'Email\s*[-–:]\s*([^\s\n]+@[^\s\n]+)', content)
        if email_match:
            patient['email'] = email_match.group(1).strip()

        # Phone
        phone_match = re.search(r'Phone\s*[-–:]\s*([^\n]+)', content)
        if phone_match:
            patient['phone'] = phone_match.group(1).strip()

        return patient

    def extract_section(self, content: str, pattern: str, multiline: bool = False) -> Optional[str]:
        """Extract a section using regex pattern"""
        flags = re.IGNORECASE | re.DOTALL if multiline else re.IGNORECASE
        match = re.search(pattern, content, flags)
        return match.group(0) if match else None

    def parse_consultation_info(self, content: str) -> Dict[str, Any]:
        """Extract consultation and medical history information"""
        consultation = {}

        # Present complaint
        complaint_match = re.search(r'Present complaint\*\*\s*\n(.*?)(?=\*\*|No Covid|TM|$)', content, re.DOTALL)
        if complaint_match:
            consultation['presentComplaint'] = self.clean_text(complaint_match.group(1))

        # COVID history
        covid_matches = re.findall(r'(No Covid[^*\n]*|Covid positive[^*\n]*|No vaccine[^*\n]*|vaccine[^*\n]*)', content, re.IGNORECASE)
        if covid_matches:
            consultation['covidHistory'] = ', '.join(covid_matches)

        # TM practice
        tm_match = re.search(r'TM.*?(?=\n|$)', content)
        if tm_match:
            consultation['tmPractice'] = tm_match.group(0)

        return consultation

    def parse_assessment(self, content: str) -> Dict[str, Any]:
        """Extract clinical assessment information"""
        assessment = {
            'vitals': {},
            'lifestyle': {'food': {}},
            'ayurvedicAssessment': {}
        }

        # Vitals
        vitals_patterns = {
            'pulse': r'Pulse\s*[-–]\s*([^\n]+)',
            'bowel': r'Bowel\s*[-–]\s*([^\n]+)',
            'urination': r'Urination\s*[-–]\s*([^\n]+)',
            'tongue': r'Tongue\s*[-–]\s*([^\n]+)',
            'sleep': r'Sleep\s*[-–]\s*([^\n]+)',
            'hunger': r'Hunger\s*[-–]\s*([^\n]+)',
            'thirst': r'Thirst\s*[-–]\s*([^\n]+)',
            'menstruation': r'Menstruation\s*[-–]\s*([^\n]+)'
        }

        for key, pattern in vitals_patterns.items():
            match = re.search(pattern, content, re.IGNORECASE)
            if match:
                assessment['vitals'][key] = match.group(1).strip()

        # Lifestyle
        lifestyle_patterns = {
            'dailyRoutine': r'Daily routine\s*[-–]\s*([^\n]+)',
            'smoking': r'Smoking\s*[-–]\s*([^\n]+)',
            'alcohol': r'Alcohol\s*[-–]\s*([^\n]+)',
            'exercise': r'Exercise\s*[-–]\s*([^\n]+)',
            'emotions': r'Emotions\s*[-–]\s*([^\n]+)'
        }

        for key, pattern in lifestyle_patterns.items():
            match = re.search(pattern, content, re.IGNORECASE)
            if match:
                assessment['lifestyle'][key] = match.group(1).strip()

        # Food
        food_patterns = {
            'hoteling': r'Food\s*[-–]\s*([^\n]+)',
            'oils': r'Oil\s*[-–]\s*([^\n]+)',
            'breakfast': r'Breakfast\s*[-–]\s*([^\n]+)',
            'lunch': r'Lunch\s*[-–]\s*([^\n]+)',
            'dinner': r'Dinner\s*[-–]\s*([^\n]+)',
            'fruits': r'Fruits\s*[-–]\s*([^\n]+)'
        }

        for key, pattern in food_patterns.items():
            match = re.search(pattern, content, re.IGNORECASE)
            if match:
                assessment['lifestyle']['food'][key] = match.group(1).strip()

        # Ayurvedic Assessment
        ayur_patterns = {
            'prakruti': r'Prakruti\s*[-–]\s*([^\n]+)',
            'vikruti': r'Vikruti\s*[-–]\s*([^\n]+)',
            'dosha': r'Dosha\s*[-–]\s*([^\n]+)',
            'dushya': r'Dushya\s*[-–]\s*([^\n]+)'
        }

        for key, pattern in ayur_patterns.items():
            match = re.search(pattern, content, re.IGNORECASE)
            if match:
                assessment['ayurvedicAssessment'][key] = match.group(1).strip()

        return assessment

    def parse_treatment(self, content: str) -> Dict[str, Any]:
        """Extract treatment recommendations"""
        treatment = {
            'shamana': [],
            'panchakarma': [],
            'additionalAdvice': []
        }

        # Shamana (Internal Medicine)
        shamana_section = self.extract_section(content, r'Shamana.*?(?=Panchakarma|Advice|Avoid)', multiline=True)
        if shamana_section:
            medicine_lines = re.findall(r'(\d+)\.\s*\*\*(.*?)\*\*', shamana_section)
            for _, medicine_text in medicine_lines:
                # Parse medicine and dosage
                if '–' in medicine_text:
                    parts = medicine_text.split('–', 1)
                    medicine_name = parts[0].strip()
                    dosage = parts[1].strip() if len(parts) > 1 else ""
                    treatment['shamana'].append({
                        'medicine': medicine_name,
                        'dosage': dosage
                    })

        # Panchakarma
        panchakarma_section = self.extract_section(content, r'Panchakarma\*\*(.*?)(?=\*\*Advice|\*\*Avoid)', multiline=True)
        if panchakarma_section:
            panchakarma_lines = re.findall(r'(\d+)\.\s*\*\*(.*?)\*\*', panchakarma_section)
            for _, therapy in panchakarma_lines:
                treatment['panchakarma'].append(therapy.strip())

        # Additional Advice (if any specific advice sections exist)
        advice_patterns = [
            r'Aroma therapy[^\n]*',
            r'Chanting of mantras[^\n]*',
            r'Advice to do Moolabandha',
            r'Use of Asafetida[^\n]*',
            r'Drink medicated water[^\n]*'
        ]

        for pattern in advice_patterns:
            matches = re.findall(pattern, content, re.IGNORECASE)
            treatment['additionalAdvice'].extend(matches)

        return treatment

    def clean_text(self, text: str) -> str:
        """Clean text by removing extra formatting and whitespace"""
        # Remove markdown formatting
        text = re.sub(r'\*\*', '', text)
        # Remove extra whitespace
        text = re.sub(r'\s+', ' ', text)
        return text.strip()

    def parse_file(self, filepath: str) -> Dict[str, Any]:
        """Parse a single consultation file"""
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()

        consultation_data = {
            'patient': self.parse_patient_info(content),
            'consultation': self.parse_consultation_info(content),
            'assessment': self.parse_assessment(content),
            'treatment': self.parse_treatment(content),
            'standardRecommendations': self.standard_recs,
            'metadata': {
                'createdAt': datetime.now().isoformat(),
                'lastModified': datetime.now().isoformat(),
                'version': '1.0',
                'language': 'en',
                'originalFile': os.path.basename(filepath)
            }
        }

        return consultation_data

    def convert_all_files(self, input_dir: str = '.', output_dir: str = './converted') -> None:
        """Convert all consultation markdown files to JSON"""
        os.makedirs(output_dir, exist_ok=True)

        # Find all consultation files
        consultation_files = [f for f in os.listdir(input_dir)
                            if f.endswith('.md') and
                            'consultation' not in f.lower() and
                            'template' not in f.lower() and
                            'readme' not in f.lower()]

        print(f"Found {len(consultation_files)} consultation files to convert")

        for filename in consultation_files:
            try:
                print(f"Converting {filename}...")
                filepath = os.path.join(input_dir, filename)
                consultation_data = self.parse_file(filepath)

                # Create output filename
                base_name = os.path.splitext(filename)[0]
                output_filename = f"{base_name}.json"
                output_path = os.path.join(output_dir, output_filename)

                # Save JSON
                with open(output_path, 'w', encoding='utf-8') as f:
                    json.dump(consultation_data, f, indent=2, ensure_ascii=False)

                print(f"✓ Converted to {output_filename}")

            except Exception as e:
                print(f"✗ Error converting {filename}: {str(e)}")

        print(f"\nConversion complete! Check the '{output_dir}' directory for results.")

def main():
    parser = ConsultationParser()
    parser.convert_all_files()

if __name__ == "__main__":
    main()